include/btllib/mi_bloom_filter.hpp

```diff
@@ -104,787 +76,600 @@ class MIBloomFilter
+  bv_rank_support = sdsl::rank_support_il<1>(&il_bit_vector);
+
+  // init counts array
+  counts_array = std::unique_ptr<std::atomic<uint16_t>[]>(
+    new std::atomic<uint16_t>[id_array_size]);
+  std::memset(
+    (void*)counts_array.get(), 0, id_array_size * sizeof(counts_array[0]));
+
+  log_info(
+    "MIBloomFilter: Bit vector size: " + std::to_string(il_bit_vector.size()) +
+    "\nPopcount: " + std::to_string(get_pop_cnt()));
+}
 
-  std::vector<T> get_data(const std::vector<uint64_t>& rank_pos) const
-  {
-    std::vector<T> results(rank_pos.size());
-    for (unsigned i = 0; i < m_hash_num; ++i) {
-      results[i] = m_data[rank_pos[i]];
-    }
-    return results;
-  }
+template<typename T>
+inline MIBloomFilter<T>::MIBloomFilter(size_t bv_size,
+                                       unsigned hash_num,
+                                       std::string hash_fn)
+  : bv_size(bv_size)
+  , hash_num(hash_num)
+  , hash_fn(std::move(hash_fn))
+{
+  bit_vector = sdsl::bit_vector(bv_size);
+}
 
-  T get_data(uint64_t rank) const
-  {
-    return m_data[rank];
-  }
+template<typename T>
+inline MIBloomFilter<T>::MIBloomFilter(sdsl::bit_vector& bit_vector,
+                                       unsigned hash_num,
+                                       std::string hash_fn)
+  : bit_vector(bit_vector)
+  , hash_num(hash_num)
+  , hash_fn(std::move(hash_fn))
+{
+  complete_bv_insertion();
+}
 
-  /*
-   * Preconditions:
-   * 	frame_probs but be equal in size to multiMatchProbs
-   * 	frame_probs must be preallocated to correct size (number of ids + 1)
-   * Max value is the largest value seen in your set of possible values
-   * Returns proportion of saturated elements relative to all elements
-   */
-  double calc_frame_probs(std::vector<double>& frame_probs,
-                          unsigned allowed_miss)
-  {
-    double occupancy = double(get_pop()) / double(size());
-    std::vector<size_t> count_table =
-      std::vector<size_t>(frame_probs.size(), 0);
-    double sat_prop = double(get_id_counts(count_table));
-    size_t sum = 0;
-    for (size_t i = 1; i < count_table.size(); ++i) {
-      sum += count_table[i];
-    }
-    sat_prop /= double(sum);
-    for (size_t i = 1; i < count_table.size(); ++i) {
-      frame_probs[i] =
-        calc_prob_single_frame(occupancy,
-                               m_hash_num,
-                               double(count_table[i]) / double(sum),
-                               allowed_miss);
+template<typename T>
+inline void
+MIBloomFilter<T>::insert_bv(const uint64_t* hashes)
+{
+  assert(!bv_insertion_completed);
+  // check array size = hash_num
+  for (unsigned i = 0; i < hash_num; ++i) {
+    uint64_t pos = hashes[i] % bit_vector.size();
+    uint64_t* data_index = bit_vector.data() + (pos >> 6); // NOLINT
+    uint64_t bit_mask_value = (uint64_t)1 << (pos & 0x3F); // NOLINT
+    (void)(__sync_fetch_and_or(data_index, bit_mask_value) >>
```

Comment:
The bit vector is from SDSL library and it is not replaceable with atomic arrays at the moment. Because that bit vector is later interleaved in the miBF construction stage and that is handled by SDSL.